rfcs/004 - Better Policy UX RFC.md

```diff
@@ -0,0 +1,156 @@
+a way to construct IAM policies in a more user-friendly way. For example, the customer intends to create an S3 bucket with
+read-only permissions from all Principals within the same AWS account. Ideally, something like this:
+
+```json
+{
+  "Version": "2012-10-17",
+  "Statement": [
+    {
+      "Effect": "Allow",
+      "Resource": "aws:arn:123456789012:bucket-resource",
+      "Action": [
+        "s3:GetObjects"
+      ],
+      "Principal": {
+        "AWS": "arn:aws:iam::123456789012:root"
+      }
+    }
+  ]
+}
+```
+
+Instead of the customer having to look up and put together a JSON blob like this and then verifying and then inserting, we can also give them
+the option to do this using functions that would generate this blob.
+
+```sql
+-- User creates policy
+select * from iasql_create_policy(
+    'some_policy_name',
+    '2012-10-17' -- nullable, defaults to 2012-10-17
+    );
+
+-- User creates statement and attaches to the policy
+select * from iasql_create_policy_statement(
+        '<policy_id>',
+        'allow',
+        'aws:arn:123456789012:bucket-resource',
+        's3:GetObjects', -- comma separated list of actions
+        '123456789012' -- comma separated list of principals
+    );
+
+-- User creates bucket with the policy
+insert into bucket (
+                name,
+                policy_document, -- nullable if policy_id is provided
+                policy_id -- nullable if policy_document is provided
+            ) values ('some_bucket', NULL, '<policy_id>'); -- triggers a function that would generate the policy document, given the policy_id and store in the policy_document field
+```
+
+Perhaps these functions could even be used together like so
+    
+```sql
+-- User creates bucket with the policy
+insert into bucket (
+                name,
+                policy_document, -- nullable if policy_id is provided
+                policy_id -- nullable if policy_document is provided
+            ) values (
+                      'some_bucket',
+                      NULL,
+                      iasql_create_policy( -- returns policy ID after creation
+                          'some_policy_name',
+                          '2012-10-17', -- nullable, defaults to 2012-10-17
+                          iasql_create_policy_statement -- returns policy statement ID after creation
+                            '<policy_id>',
+                            'allow',
+                            'aws:arn:123456789012:bucket-resource',
+                            's3:GetObjects', -- comma separated list of actions
+                            '123456789012' -- comma separated list of principals
+                            ), -- Statement 1
+                          iasql_create_policy_statement(...), -- Statement 2
+                          ...
+                          iasql_create_policy_statement(...) -- Statement N
+                    );
```

Comment:
If I replace this insert with a JSON string, even if human-legibly formatted, it seems to have no impact on complexity for the user?

```suggestion
insert into bucket (
                name,
                policy_document, -- nullable if policy_id is provided
                policy_id -- nullable if policy_document is provided
            ) values (
                      'some_bucket',
                      $${
                        "Version": "2012-10-17",
                        "Statement": [{
                          "Effect": "Allow",
                          "Action": [ "s3:GetObjects" ],
                          "Resource": "aws:arn:123456789012:bucket-resource",
                          "Prinicpal": {
                            "wheredidthiskeygo": "123456789012"
                          }
                        }]
                      }$$
            );
```

So I don't think it is providing what could be considered a better experience for the users?